More click modes (with tests!)

kaduvert opened the pull request:
Hey there,

this pull request implements:

shift clicks,
number clicks,
middle clicks
and drop clicks

or mode 1-4 as seen in

![this image](https://user-images.githubusercontent.com/70949550/188121401-baeb8d4b-d19e-497d-b043-6858b14a6422.png)

with typescript definitions, **tests (mocha)** and improved code readability.

You may recall me from a pull request one and a half years ago, which closed due to lack of tests and maybe other stuff.
I finally got around to working on it and i think i've addressed all the issues! :)

kaduvert committed:
initial commit

kaduvert committed:
added mocha tests

kaduvert committed:
updated typescript typings to match main repo

kaduvert committed:
improved on readability

kaduvert committed:
improved on typescript definitions

rom1504 commented on index.js:
```diff
@@ -57,7 +57,7 @@ function loader (mcVersion) {
 
   const windowByType = new Map()
-  for (const key of Object.keys(windows)) {
+  for (const key in windows) {
```
don't use in, in javascript it's bad
prefer for or .forEach

kaduvert replied:
sorry for that one, removed.

rom1504 commented:
looks ok

need more review

could you check @u9g maybe ?

kaduvert committed:
Merge branch 'master' into master

kaduvert committed:
readded mocha to dependencies after accidentally removing it through resolving a merge conflict

kaduvert committed:
removed unnecessary change

kaduvert committed:
simplified one function for improved readability

u9g commented on index.d.ts:
```diff
@@ -132,6 +218,13 @@ export class Window<T = unknown> extends (EventEmitter as new <T>() => TypedEmit
+     * @param start
+     * @param end
+     */
+    sumRange(start: number, end: number): number;
```
can you give a bit more info on what this is counting? what are start and end?

kaduvert replied:
Returns how much items there are ignoring what the item is

basically taking start slot and end slot and counting the items from start to end slot, but i think just reading the function code would make this obvious

edit: i clarified that in the typescript definitions a bit

u9g commented on index.js:
```diff
@@ -67,8 +67,7 @@ function loader (mcVersion) {
     createWindow: (id, type, title, slotCount = undefined) => {
-      let winData = windowByType.get(type)
-      if (!winData) winData = windows[type]
+      let winData = windowByType.get(type) || windows[type]
```
```suggestion
      let winData = windowByType.get(type) ?? windows[type]
```

u9g commented on lib/Window.js:
```diff
@@ -23,105 +23,204 @@ module.exports = (Item) => {
-      assert.ok(click.mouseButton === 0 || click.mouseButton === 1)
+      const { mode, slot, mouseButton } = click
+      assert.ok(
+        (mode <= 6 && mode >= 0) &&
```
```suggestion
        (mode >= 0 && mode <= 6) &&
```

u9g commented on lib/Window.js:
```diff
@@ -23,105 +23,204 @@ module.exports = (Item) => {
+          break
+
+        case 2:
+          assert.ok(mouseButton <= 8, 'invalid operation')
```
You're ignoring mode 2, button 40, which is offhand swap key

kaduvert replied:
in minecraft 1.8 there is no offhand,
i just implemented clicks for 1.8, as i have no experience with minecraft above version 1.8 and don't feel confident implementing that

IceTank replied:
Do you want to add offhand swap? It does not have to be in this pr but supporting it at some point would be nice.

IceTank replied:
I don't consider this as a blocker

u9g commented on lib/Window.js:
```diff
@@ -23,105 +23,204 @@ module.exports = (Item) => {
+          break
+
+        case 5:
+          this.dragClick(click)
```
if you're doing asserts on the rest, might as well assert that this is in [1,5,9,2,6,10]

u9g commented on lib/Window.js:
```diff
@@ -23,105 +23,204 @@ module.exports = (Item) => {
+        this.dropSelectedItem(click.mouseButton === 0)
       } else {
-        return this.acceptNonInventorySwapAreaClick(click)
+        let { item } = click
```
why were the smaller methods removed?

kaduvert replied:
they were hard to read and inflexible.

i read the entire old code and figured a complete rewrite of the functions directly handling clicks is the best option.

u9g reviewed (changes requested):
Some thoughts, I'll finish the review after these get fixed.

kaduvert committed:
clarified what sumRange() does

kaduvert committed:
implement u9g's suggestion

kaduvert committed:
improved variable naming

kaduvert committed:
implement u9g's suggestion

kaduvert committed:
implement u9g's suggestion

kaduvert committed:
lint fix

kaduvert committed:
Merge branch 'master' into master

kaduvert committed:
made compatible with prismarine-registry refactoring

kaduvert committed:
small fix

KadaverBrutalo10 commented:
@u9g

IceTank commented on index.js:
```diff
@@ -1,7 +1,7 @@
 function loader (registryOrVersion) {
   const registry = typeof registryOrVersion === 'string' ? require('prismarine-registry')(registryOrVersion) : registryOrVersion
-  const Item = require('prismarine-item')(registry)
```
Why was this changed? It is more efficient if prismarine-item can use the same registry version as p window.

kaduvert replied:
because of https://github.com/PrismarineJS/prismarine-windows/pull/77

i tested my pr with these changes and my tests failed,
i quickly looked up how to get the mcVersion and implemented that so everything worked again.

i don't know anything about this refactoring i just made my pr compatible with it
if you want you can edit it, but make sure the tests all succeed

IceTank replied:
I will look at it. But I don't see a reason why that would fail. Maybe you where not using up to date dependencies?

kaduvert replied:
i think it was because prismarine-item wanted a mcversion string instead of some registry object

kaduvert replied:
fixed.

IceTank commented on lib/Window.js:
```diff
@@ -23,105 +23,205 @@ module.exports = (Item) => {
       }
     }
 
-    acceptOutsideWindowClick (click) {
```
Changing all the already implemented function to new functions with different names is a breaking change. We won't be able to release this without a major version jump or this will break everything that is using prismarine-windows.

kaduvert replied:
As stated above "they were hard to read and inflexible"

The functions that were removed should only be called by acceptClick() and are probably also not called from anywhere outside, as the removed functions simply split up the logic of acceptClick()

IceTank commented:
Are new function implementation really necessary? If prismarine-windows is changing everything we also have to re implement windows in mineflayer.

kaduvert commented:
@IceTank Yes, i think so.
The old functions were unworkable with,
that's the reason i removed them and likely also the reason why nobody else implemented other click types for a long time

Also this doesn't break mineflayer, if mineflayer was to use this then it would work as before, because the only clicks currently allowed by mineflayer are normal ones (mode 0) and they work as they did before

rom1504 commented:
@u9g @IceTank @KadaverBrutalo10 thoughts: ready to merge ?

kaduvert commented:
@rom1504 sure

IceTank commented:
Still looks like this is a breaking change in the API. There might be some projects that break even with the small API changes.
This does looks like a good change tho and we should definitely add more click options.
But, as I understand, this was only made and tested for version 1.8. We need to test how well it works for newer versions too. 
Maybe someone can do a bigger manual test? We could just jump head first into it and try to patch stuff after but I have a feeling that this could break a lot of things in mineflayer.

KadaverBrutalo10 commented:
Well you see, for projects to break, projects first have to depend on the removed functionality in the first place.
Just to make sure we're talking about the same thing, these are the functions that would be removed:

```js
acceptOutsideWindowClick(click)
acceptInventoryClick(click)
acceptNonInventorySwapAreaClick(click)
acceptSwapAreaRightClick(click)
acceptSwapAreaLeftClick(click)
```

These are all functions, that are called within prismarine-windows, and within prismarine-windows only.
The purpose of `acceptClick(click)` is, to split up internal click-handling logic.
Code outside of prismarine-windows **should not call** these functions.

The only function it makes sense to call from outside is `acceptClick(click)` to accept a click of any kind.

So if any projects depend on these functions:
first of all, they shouldn't, but
second of all, they can just call `acceptClick(click)` instead! (they should be doing this already)
Currently, **not even mineflayer (v4.8.1)** calls *any* of the removed functions directly.

As to the testing side of things, the currently supported click modes (just 0) **work as they did before**, because the logic remains unchanged. Sure, the function names are different and all that (for improved read- and reuse-ability), but the raw logic remains the same.

Because things are taking long (~1.5 yrs) i did extensive testing of my changes already through normal use.
Over time, others even *starred and forked* my fork, so it doesn't seem to be so bad after all...

Why don't you go ahead and test/review my changes? Why aren't you the one doing the "bigger manual tests"?

kaduvert committed:
add gamemode arg to functions that (might) need it

kaduvert committed:
implemented middleClick

kaduvert committed:
implemented tests for middleClick

kaduvert committed:
fixed click code order to be ascending from 0-6

kaduvert committed:
removed gamemode arg from dropClick in index.d.ts (fix)

KadaverBrutalo10 commented:
sorry @IceTank if my last comment sounded a bit aggressive. :(
i likely even commented on statements, that weren't even directed at me.
i just wanted to make it clear, that my changes are not causing major changes to mineflayer and also not to any realistic projects using prismarine-windows.

to make up for that (and maybe also because i need it), i implemented middleClicks :)

kaduvert committed:
resolve deprecation concerns once and for all

kaduvert committed:
fix lint (duplicate declaration)

kaduvert commented:
forget what i said about how unreasonable it would be to call the removed functions, i added them back in.

they now are all aliases of acceptClick... and yes, this is perfectly fine.
hope everyone's happy now and merging can begin! :)

kaduvert committed:
clarify gamemode argument in definitions

IceTank commented on test/test.js:
```diff
@@ -0,0 +1,141 @@
+const mocha = require('mocha')
+const { describe, it } = mocha
```
```suggestion
const { describe, it } = require('mocha')
```

kaduvert replied:
implemented.

IceTank commented on test/test.js:
```diff
@@ -0,0 +1,141 @@
+
+const assert = require('assert')
+
+const windows = require('..')('1.8')
```
Have you tested this with other versions as well?

kaduvert replied:
no

to my knowledge, the click modes implemented here have not changed much since 1.8

IceTank commented on test/test.js:
```diff
@@ -0,0 +1,141 @@
+  item: slot === -999 ? null : useInv ? inv.slots[slot] : chest.slots[slot]
+})
+
+chest.updateSlot(0, new Item(1, 64))
```
Please don't use hard coded item id's. I know this is a 1.8 specific test but item id's change between versions. It is better to use `mcData.itemsByName.dirt.id` for instance.

kaduvert replied:
implemented.

IceTank commented on test/test.js:
```diff
@@ -0,0 +1,141 @@
+    })
+
+    it.skip('shift out of inventory into armor slot - leather boots', () => {
+      // isn't even implemented, so it would naturally fail.
```
Is the implementation planed? Or is there an active issue for this?

kaduvert replied:
neither,
there is no issue because in the current prismarine-windows you'd have to explicitly trigger this behaviour.
as to implementing, i couldn't really figure out how to do it, as i didn't find any mcData stuff which describes which items can go into which slot

an issue should probably be created for this.

IceTank commented on test/test.js:
```diff
@@ -0,0 +1,141 @@
+chest.updateSlot(chest.inventoryEnd - 3, new Item(1, 44))
+
+inv.updateSlot(inv.inventoryEnd - 1, new Item(301, 1))
+inv.updateSlot(inv.inventoryEnd - 2, new Item(1, 1))
```
So the chest content and inventory content is initialized as the test is loaded. What happens when clicks are accepted? Do the items move inside of the container? Maybe it is better to use mocha's `beforeEach()` hook? https://mochajs.org/#hooks

You can run specific tests with mocha by using regex to match the test name. Do the tests break if they are not all executed after each other?

kaduvert replied:
yes, the tests absolutely suck and will break if they are not executed in order xD

i have no idea how mocha works, just wanted to make minimal tests, which show that my changes work.
if you understand the logic behind my tests, you are likely able to improve them by a lot

IceTank commented on package.json:
```diff
@@ -4,7 +4,7 @@
   "main": "index.js",
   "scripts": {
-    "test": "npm run lint",
+    "test": "npm run lint && mocha",
```
Test's in prismarine projects are usually implemented with scripts that look like this 
```json
"mocha_test": "mocha --reporter spec --exit",
"test": "npm run mocha_test",
```
See mineflayers package.json file https://github.com/PrismarineJS/mineflayer/blob/master/package.json

kaduvert replied:
implemented.

kaduvert committed:
IceTank's require suggestion

kaduvert committed:
tests now use itemsByName

kaduvert committed:
implement npm tests in package.json as suggested

IceTank commented on test/test.js:
```diff
@@ -0,0 +1,145 @@
+  describe('mouseButton 0', () => {
+    it('pickup item', () => {
+      chest.acceptClick(createClick(0, 0, 0))
+      assert(!chest.slots[0] && chest.selectedItem?.count === 64 && chest.selectedItem.type === 1)
```
Don't put multiple assertions into one assert. This will make a test unreadable if it does fail. The assertion is also missing a meaningful comment of what is being asserted. For instance this line should be changed to something like this.
```js
assert(!chest.slots[0], 'Slot 0 should be empty')
assert(chest.selectedItem?.count === 64 && chest.selectedItem?.type === 1, 'Selected item in chest should be stone (??) and have count 64')
```

IceTank commented on test/test.js:
```diff
@@ -0,0 +1,145 @@
+  describe('mouseButton 0', () => {
+    it('pickup item', () => {
+      chest.acceptClick(createClick(0, 0, 0))
+      assert(!chest.slots[0] && chest.selectedItem?.count === 64 && chest.selectedItem.type === 1)
```
Please use the stoneId variable

kaduvert replied:
oops, didn't notice that id's are also used elsewhere.

will fix

IceTank commented on test/test.js:
```diff
@@ -0,0 +1,145 @@
+
+    it('number click from slot with item to slot with different item', () => {
+      chest.acceptClick(createClick(2, 5, 5))
+      assert(!chest.slots[5] && chest.slots[59].type === 2 && chest.slots[59]?.count === 3 && chest.slots[54]?.count === 4 && chest.slots[54].type === 1)
```
```suggestion
      assert(!chest.slots[5] && chest.slots[59]?.type === 2 && chest.slots[59]?.count === 3 && chest.slots[54]?.count === 4 && chest.slots[54].type === 1)
```

IceTank commented on test/test.js:
```diff
@@ -0,0 +1,145 @@
+
+    it('get stack into selectedItem (gamemode 1)', () => {
+      inv.acceptClick(createClick(3, 2, inv.inventoryEnd - 2, true), 1)
+      assert(Item.equal(inv.selectedItem, new Item(1, inv.slots[inv.inventoryEnd - 2].stackSize)))
```
```suggestion
      assert(Item.equal(inv.selectedItem, new Item(1, inv.slots[inv.inventoryEnd - 2]?.stackSize)))
```

IceTank commented:
Tests work for version 1.8 but every other version does not work. Have not tested on a real server.

rom1504 commented:
hey @kaduvert would you like to finish this ?

supporting more clicks seems like a great add

kaduvert committed:
completely rewrote tests

kaduvert committed:
removed unnecessary complexity

kaduvert committed:
added number click (mode 2) testing for 1.9 and higher

kaduvert committed:
use registry for version determining

kaduvert committed:
small fix

kaduvert commented:
sorry for the delay guys.
this should fix everything bad with the old tests in terms of readability and it eliminates dependency on previously executed tests.

kaduvert committed:
added missing assertion

kaduvert committed:
variable renaming

kaduvert committed:
added eslint-env mocha comment to tests

kaduvert committed:
more renaming for better readability

kaduvert committed:
added detection of changed, but unasserted slots

kaduvert committed:
small fix

kaduvert committed:
added assertions for unasserted, but updated slots

kaduvert committed:
improved error message

kaduvert commented:
i now also added checks for updated but unasserted slots, meaning that if you test a click and it updates slots, you must assert all the updatedSlots or it fails.
this should help detect odd / unexpected behaviour.

i also fixed the version issue, tests now work with versions above 1.8.

kaduvert committed:
small fix

kaduvert committed:
improved error messages

kaduvert committed:
misc fix: updated prismarine-item to take registry object

kaduvert committed:
small fix: tests now also share registry

kaduvert committed:
renamed asserting functions to better express intent

IceTank commented on index.d.ts:
```diff
@@ -49,24 +49,118 @@ export class Window<T = unknown> extends (EventEmitter as new <T>() => TypedEmit
 
-    acceptClick(click: Click): void;
+    /**
+     * deprecated, call acceptClick(click) instead
```
A bit of a nit pick: There is a `@deprecated` jsdoc tag that strikes through a function when used in an ide. https://jsdoc.app/tags-deprecated.html. For this to work correctly the tag has to be in the index.d.ts file as this is where the ide's take there IntelliSense from and it has to be annotate for every function. A compact inline way to achieve this is by writing it like this
```js
/** @deprecated use {@link acceptClick} instead */
acceptNonInventorySwapAreaClick
```
the `@link` makes the linked function clickable in the IntelliSense preview.

IceTank replied:
The rest looks good to me. I have not tested it on an actual server and I won't be able to for some time. But if tests pass I think we can go ahead and merge it. We will just have to see if there are issues when people start to complain.

kaduvert committed:
Merge branch 'master' into master

IceTank commented on lib/Window.js:
```diff
@@ -22,106 +22,219 @@ module.exports = (Item) => {
+          break
+
+        case 2:
+          assert.ok(mouseButton <= 8, 'invalid operation')
```
It looks like mouseButton for above 8 is possible in at least 1.19.4 (Although not vanilla) https://web.archive.org/web/20230519040131/https://wiki.vg/Protocol#Click_Container (web archive because I can't seam to access wiki.vg). So some sort of switch for newer versions would be nice at some point.

kaduvert committed:
turned deprecation comment into proper jsdoc

kaduvert committed:
fixed whitespace in ts doc

kaduvert commented:
so i see that mouseButtons higher than 8 are a thing in newer versions and even 40 for offhand, but i have no clue and i can not confidently implement it because i have almost no experience with minecraft above 1.8.
also, one has to keep in mind, that the offhand slot is picky with what it accepts.

kaduvert commented:
let's not forget that the click modes implemented here already come with huge time-benefits in window management.

all this other stuff like mode 5 clicks or mouseButtons higher than 8 in mode 2 are definitely useful, but i think that you're getting to the point of diminishing returns implementing them. also, i personally neither use or need the unimplemented stuff currently.

the code is structured so that you can add new modes, like 5 and 6 without much friction, so anyone who wants to can do it easily.

kaduvert commented:
prismarine-windows as a whole has not implemented any logic for handling special slots, like armor slots in your inventory or the bottom slots in the brewing stand, which only accept some items.

this is not a huge problem with click modes 0 and 2-6.
you **could** in theory desync your window from the server one by moving the wrong items into slots which don't accept it, but you'd have to explicitly trigger this behaviour by specifying the slot as some kind of target. so it's unlikely to happen to anyone by accident.

with mode 1 on the other hand, this behaviour **can** happen without intent, for example if you shift around a piece of armor in an inventory, on the server it will go into the respective armor slot (if it's free), but because the client does not handle these special slots, it will (if the slot was free) lead to desync.

kaduvert commented:
so what i suggest for now is labeling shift clicks as unstable in inventorys that have special slots until proper handling is implemented.
I could implement this handling, but i don't know where to find the data for which slots accepts which types of item, or if this data even exists. If it doesn't, i also wouldn't know how to reliably produce it, especially because this potentially changes with every minecraft version.

other than that however, there is no issues i know of.

IceTank commented:
> also, one has to keep in mind, that the offhand slot is picky with what it accepts

This is only true for bedrock. In Java the offhand slot is like any other inventory slot and accepts all items.

IceTank commented:
If the user tries to shit click an item into a slot he is not suppose to I would consider it user error. Also there is a good chance the server just rejects the click resulting in no change in the inventory.

kaduvert commented:
Whatever, but i think it should be noted somewhere in the docs, just so the user knows.
Also, if the server rejects, the client immediately exits with an error https://github.com/PrismarineJS/mineflayer/blob/f50d904fca965ad9aef06ce31b7acfa1a7e8baa9/lib/plugins/inventory.js#LL549C1-L555C8

As to offhand, i just don't know.
someone else probably knows it better, so i will not try to implement it.
If you really want to, i can try, but i'd rather not.

IceTank commented:
> Also, if the server rejects, the client immediately exits with an error

This is not good. Those rejections happen a lot and should be catchable. 

> https://github.com/PrismarineJS/mineflayer/blob/f50d904fca965ad9aef06ce31b7acfa1a7e8baa9/lib/plugins/inventory.js#LL549C1-L555C8

This peace of code is truly awful and should be refactored. But this is a concern for mineflayer not prismarine windows.

The offhand does accept all items. You can test it with a vanilla client in single player. But implementing that feature is not a priority I would say. We can add it at some later date.

kaduvert commented:
Yeah, you're right with the offhand, i just checked.

Anyways, are there any more issues or can this be merged?

kaduvert committed:
added left-out mouseButton checks for compatibility / readability

rom1504 commented:
Looks pretty good to me.

I think maybe one thing to try is open a PR in mineflayer using this to make sure it breaks no test.

Besides that I think I'll merge in next days if nobody has something to say

kaduvert committed:
include offhand slot in assert to not refuse valid clicks

kaduvert committed:
added returning of changedSlots for mc versions >= 1.17

kaduvert committed:
replaced structuredClone with JSON reparsing for compatibility

kaduvert commented:
did it: https://github.com/PrismarineJS/mineflayer/pull/3102

all tests green @rom1504

rom1504 commented:
thanks for all the work on this

rom1504 commented:
/makerelease 2.7.0
